### ISSUE TITLE
Index Factory IVF_HNSW metric

### ISSUE THREAD

**Issue by psj8252** [facebookresearch/faiss#1316]
https://github.com/facebookresearch/faiss/blob/ac74f576f75089784d752543548824a555df4e6c/faiss/index_factory.cpp#L141-L145

In the code when I use IVF_HNSW index, the metric should be METRIC_L2.
But in my testing, I can use HNSWFlat with METRICS_INNER_PRODUCT as a quantizer of IVFPQ Index.
```cpp
  auto quantizer = faiss::IndexHNSWFlat(dimension, M, faiss::METRIC_INNER_PRODUCT);
  auto index = faiss::IndexIVFPQ(&quantizer, dimension, nlist, pq_m, 8, faiss::METRIC_INNER_PRODUCT);
```
And this worked.

So I wonder why there is an assertion that ensure the mertric is L2.

**Comment by mdouze** [comment]
This is a trace of when HSNW did not support inner product as a metric. So it should indeed be removed.

**Comment by pandasMX** [comment]
Hi, i have removed the wrong assertion in index_factory.cpp and try to train the index on 
`faiss.index_factory(d, 'OPQ32_128,IVF4096_HNSW32,PQ32', faiss.METRIC_INNER_PRODUCT)`, 
but when the training finished, it threw another error
`RuntimeError: Error in void faiss::Level1Quantizer::train_q1(size_t, const float*, bool, faiss::MetricType) at IndexIVF.cpp:93: Error: 'metric_type == METRIC_L2' failed`
think we also need to remove this assertion?

**Comment by mdouze** [comment]
Fixed in PR #1555 with will be merged shortly.
